### ISSUE TITLE
[analyzer] add a fix for `uri_does_not_exist`

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#48760]
Should there be a fix for package import of [uri_does_not_exist](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=uri_does_not_exist#uri_does_not_exist), that possibly adds the suggested dependency, for example for

```dart
import 'package:web_socket_channel/web_socket_channel.dart';
```

I would imagine something like:
- adding the dependency in `pubspec.yaml` as `any` version (or getting the latest one from `pub.dev`)
- running `pub get`

**Comment by bwilkerson** [comment]
@jonasfj @mit-mit We've had discussions about this internally to explore just how much of the process we can automate. I believe that we could add the dependency, but we can't currently execute anything afterwards because of what the clients support. The next question is whether we could do some of the work before we send the edits to the client.

**Comment by jonasfj** [comment]
> The next question is whether we could do some of the work before we send the edits to the client.

As I understood the discussion. The analysis_server cannot edit files on disk, nor execute commands on the file-system where the source code is stored. Or maybe I understood that wrong... or likely partly wrong :D

But then can it do things in response to an quick-fix being triggered?


----------------------

What if we just figured out the changes for `pubspec.yaml`, but we don't actually do `dart pub get`.
So users who apply the "quick fix" would get a change in `pubspec.yaml`, and then their editor will then tell them to run `dart pub get`, because `mtime(pubspec.yaml) > mtime(pubspec.lock)`.

In fact I think many editors default to running `dart pub get` after you save `pubspec.yaml`.

**Comment by bwilkerson** [comment]
> The analysis_server cannot edit files on disk, ...

That's correct. It is often the case that the content of the file has been changed in the client's edit buffer, so changing what's on disk would cause problems.

> ... nor execute commands on the file-system where the source code is stored.

It can execute some commands on the file system. It just needs to be careful about commands that are going to modify files. So it probably wouldn't be good to execute a `pub` command that updated the `pubspec.yaml` (because the user might have unsaved changes), but we could, potentially, run `pub` to find the optimal version constraints as long as doing so didn't depend on or change the state of the files on disk. I say "potentially" because there are also some performance constraints we need to account for.

> But then can it do things in response to an quick-fix being triggered?

Yes. What it does is send back to the client a set of edits that the client then applies to the content in the edit buffer.

> What if we just figured out the changes for `pubspec.yaml`, but we don't actually do `dart pub get`.

That would be ideal, yes.

As I understood the discussion, which might also have been at least partially wrong, the primary issue there is one of performance. Because of the way the clients are implemented, the server has to compute all of the edits before the list of fixes can be returned. If it takes a long time to compute one of the fixes then
- that delays the ability of the client to display any of the other fixes that might apply at the cursor location, and
- makes the server feel slow (which is a problem we need to make better rather than worse).

That doesn't mean we shouldn't do this, we just need to figure out how much time we're talking about in order to weigh the cost and benefit of adding the fix.

(A secondary issue is that this has to be resilient to errors such as not being able to reach `pub.dev` due to a lack or internet connection, etc. Not a show-stopper, but we'd need to decide whether to not offer a fix in those cases or whether to offer a less complete fix, such as by using `any` for the version constraint. Also, users might not be able to run `dart pub get` after the edit.)
